activesupport/lib/active_support/callbacks.rb

```diff
@@ -313,14 +314,17 @@ def merge_conditional_options(chain, if_option:, unless_option:)
           self.class.build chain, @filter, @kind, options
         end
 
-        def matches?(_kind, _filter)
-          @kind == _kind && filter == _filter
+        def matches?(_kind, _filter, _target = nil)
+          callback_match = @kind == _kind && filter == _filter
+          return callback_match unless _target
+
+          target == _target && callback_match
```

Comment:
Note: Could also move this logic into the Dedupe so that matches? doesn't have to be concerned with the dedupe target.

Kept here so that all "matching" logic was centralized.